pilot/pkg/model/extensions.go

```diff
@@ -127,3 +100,63 @@ func convertToWasmPluginWrapper(plugin *config.Config) *WasmPluginWrapper {
 		ExtensionConfiguration: ec,
 	}
 }
+
+func getDataSource(u *url.URL, wasmPlugin *extensions.WasmPlugin) *envoyCoreV3.AsyncDataSource {
+	if u.Scheme == fileScheme {
+		return &envoyCoreV3.AsyncDataSource{
+			Specifier: &envoyCoreV3.AsyncDataSource_Local{
+				Local: &envoyCoreV3.DataSource{
+					Specifier: &envoyCoreV3.DataSource_Filename{
+						Filename: strings.TrimPrefix(wasmPlugin.Url, "file://"),
+					},
+				},
+			},
+		}
+	}
+
+	return &envoyCoreV3.AsyncDataSource{
+		Specifier: &envoyCoreV3.AsyncDataSource_Remote{
+			Remote: &envoyCoreV3.RemoteDataSource{
+				HttpUri: &envoyCoreV3.HttpUri{
+					Uri:     u.String(),
+					Timeout: durationpb.New(30 * time.Second),
+					HttpUpstreamType: &envoyCoreV3.HttpUri_Cluster{
+						// this will be fetched by the agent anyway, so no need for a cluster
+						Cluster: "_",
+					},
+				},
+				Sha256: wasmPlugin.Sha256,
+			},
+		},
+	}
+}
+
+func getVmConfig(datasource *envoyCoreV3.AsyncDataSource, vm *extensions.VmConfig) *envoyExtensionsWasmV3.PluginConfig_VmConfig {
+	cfg := &envoyExtensionsWasmV3.PluginConfig_VmConfig{
+		VmConfig: &envoyExtensionsWasmV3.VmConfig{
+			Runtime: defaultRuntime,
+			Code:    datasource,
+		},
+	}
+
+	if vm != nil && len(vm.Env) != 0 {
+		hostEnvKeys := make([]string, 0, len(vm.Env))
+		inlineEnvs := make(map[string]string, 0)
+		for _, e := range vm.Env {
+			switch e.ValueFrom {
+			case extensions.EnvValueSource_INLINE:
+				if e.Value != "" {
```

Comment:
I think we should allow setting an env var to an empty string?

Reply:
why user need setup an empty string ENV?

Reply:
get the  existence of an env?

Reply:
updated